code/tasks/ProcessJobQueueTask.php

```diff
@@ -51,46 +123,134 @@ public function run($request) {
 				$queue = QueuedJob::LARGE;
 				break;
 			}
-			default: {
-				// leave it as whatever this queue name is
-			}
 		}
 
-		$this->writeLogLine("$datestamp Processing queue $queue");
+		return $queue;
+	}
 
-		if ($request->getVar('list')) {
-			for ($i = 1; $i  <= 3; $i++) {
-				$jobs = $service->getJobList($i);
-				$num = $jobs ? $jobs->Count() : 0;
-				$this->writeLogLine("$datestamp Found $num jobs for mode $i");
-			}
-			return;
+	/**
+	 * Write in a format expected by the output medium (CLI/HTML).
+	 *
+	 * @param string $line Line to be written out, without the newline character.
+	 * @param null|string $prefix
+	 */
+	private function writeLogLine($line, $prefix = null) {
+		if(!$prefix) {
+			$prefix = '[' . date('Y-m-d H:i:s') . '] ';
 		}
 
-		// Cleanup or restart jobs before processing
-		$service->checkJobHealth();
-		
-		/* @var $service QueuedJobService */
-		$nextJob = null;
-		
-		// see if we've got an explicit job ID, otherwise we'll just check the queue directly
-		if ($request->getVar('job') && strpos($request->getVar('job'), '-')) {
-			list($junk, $jobId) = split('-', $request->getVar('job'));
-			$nextJob = DataObject::get_by_id('QueuedJobDescriptor', $jobId);
+		if(Director::is_cli()) {
+			echo $prefix . $line . "\n";
 		} else {
-			$nextJob = $service->getNextPendingJob($queue);
+			echo Convert::raw2xml($prefix . $line) . "<br>";
 		}
+	}
 
-		if ($nextJob) {
-			$this->writeLogLine("$datestamp Running $nextJob->JobTitle and others from $queue.");
-			$service->processJobQueue($queue);
+	/**
+	 * Logs the status of the queued job descriptor.
+	 *
+	 * @param bool|null|QueuedJobDescriptor $descriptor
+	 * @param string $queue
+	 */
+	protected function logDescriptorStatus($descriptor, $queue) {
+		if(is_null($descriptor)) {
+			$this->writeLogLine('No new jobs');
+		}
+
+		if($descriptor === false) {
+			$this->writeLogLine('Job is still running on ' . $queue);
 		}
 
-		if (is_null($nextJob)) {
-			$this->writeLogLine("$datestamp No new jobs");
+		if($descriptor instanceof QueuedJobDescriptor) {
+			$this->writeLogLine('Running ' . $descriptor->JobTitle . ' and others from ' . $queue . '.');
 		}
-		if ($nextJob === false) {
-			$this->writeLogLine("$datestamp Job is still running on $queue");
+	}
+
+	/**
+	 * @param SS_HttpRequest $request
+	 */
+	protected function runWithDoormanEngine($request) {
+		// fix/prep any strange jobs!
+
+		$service = $this->getService();
+		$service->checkJobHealth();
+
+		// split jobs out into multiple tasks...
+
+		$manager = new DoormanProcessManager();
+		// $manager->setLogPath(__DIR__);
+
+		if ($this->config()->processes > 0) {
+			$rule = new InMemoryRule();
+			$rule->setProcesses($this->config()->processes);
+
+			$manager->addRule($rule);
 		}
+
+		$descriptor = $this->getNextJobDescriptorWithoutMutex($request);
+
+		while($manager->tick() || $descriptor) {
```

Comment:
I think this loop needs some code documentation. Can you describe the business logic?